crates/conf/macros/src/lib.rs

```diff
@@ -0,0 +1,217 @@
+                            .to_compile_error()
+                            .into();
+                        }
+
+                        let arg = &closure.inputs[0];
+                        let arg = match arg {
+                            Pat::Type(arg) => arg,
+                            _ => {
+                                return syn::Error::new_spanned(
+                                    attr,
+                                    "expected a closure with one argument",
+                                )
+                                .to_compile_error()
+                                .into();
+                            }
+                        };
+                        // let arg_type = &arg.ty;
+                        let arg_type = quote! { &#field_type };
+
+                        let arg_name = &arg.pat;
+                        let arg_name = quote! { #arg_name };
+
+                        let body = &closure.body;
+                        check_fn = Some(quote! {
+                            (|#arg_name: #arg_type| {
+                                #body
+                            })
+                        });
+
+                        inner_check_fns.push(check_fn)
+                    } else if attr.path.is_ident("ensure") {
+                        // the ensure attribute has 2 inputs like this: #[ensure(move_margin > 0., "`move_margin` must be positive.")]
+                        // it uses the `ensure!()` macro from the anyhow crate
+
+                        // we need to get the two inputs separated by a comma separately. the first one is a boolean expression and the second one is a string literal
+                        // convert to tuple
+                        let ensure_args = match attr.parse_args::<EnsureArgs>() {
+                            Ok(args) => args,
+                            Err(e) => {
+                                let mut error = syn::Error::new_spanned(
+                                    attr,
+                                    "expected a tuple like: `#[ensure(move_margin > 0., \"`move_margin` must be positive.\")]`",
+                                );
+                                error.combine(e);
+                                return error.to_compile_error().into();
+                            }
+                        };
+
+                        let arg_type = quote! { &#field_type };
+
+                        let condition = ensure_args.0;
+                        let string = ensure_args.1;
+
+                        check_fn = Some(quote! {
+                            (|#field_name: #arg_type| {
+                                ensure!(#condition, #string);
+                                Ok(())
+                            })
+                        });
+
+                        inner_check_fns.push(check_fn)
+                    } else if attr.path.is_ident("is_finite") {
+                        is_finite_check = !is_finite_check;
```

Comment:
So that it can be toggled.

Reply:
I don't understand the toggling. Why isn't handle this attribute independently for each field?

Reply:
In a lot of configs I foresee that there are many number values to be validated. So that it is not tedious to type out validation for each field. It's toggle so if you want to turn it on for a group of numbers then turn it off.  Have a look at the tests I wrote to see this behaviour.

Reply:
In a lot of configs I foresee that there are many number values to be validated. So that it is not tedious to type out validation for each field. It's toggle so if you want to turn it on for a group of numbers then turn it off.  Have a look at the tests I wrote to see this behaviour.

Reply:
Now I understand, but I would go the path of least surprise. Id est I would handle this independently for each field. Or is this pattern widely used / known?